cfme/common/provider.py

```diff
@@ -153,47 +151,162 @@ def create(self, cancel=False, validate_credentials=True, check_existing=False,
+                    except AttributeError:
+                        # tabs are absent in UI when there is only single (default) endpoint
+                        endp_view = self.endpoints_form(parent=main_view)
+
+                    endp_view.fill(endpoint.view_value_mapping)
+
+                    # filling credentials
+                    if hasattr(endpoint, 'credentials'):
+                        endp_view.fill(endpoint.credentials.view_value_mapping)
+
+                        if validate_credentials and hasattr(endp_view, 'validate'):
+                            # there are some endpoints which don't demand validation like
+                            #  RSA key pair
+                            endp_view.validate.click()
+
+                if cancel:
+                    created = False
+                    main_view.cancel.click()
+                    cancel_text = 'Add of Infrastructure Provider was cancelled by the user'
+                    flash.assert_message_match(cancel_text)
+                else:
+                    main_view.add.click()
+                    # todo: replace flash when main view is merged
+                    flash.assert_message_match(
+                        '{} Providers "{}" was saved'.format(self.string_name,
+                                                             self.name))
+            if validate_inventory:
+                self.validate()
 
-        return created
+            return created
 
-    def update(self, updates, cancel=False, validate_credentials=True):
+        else:
+            if check_existing and self.exists:
+                created = False
+            else:
+                created = True
+                logger.info('Setting up provider: %s', self.key)
+                navigate_to(self, 'Add')
+                fill(self.properties_form, self._form_mapping(True, hawkular=False,
+                                                              **self.__dict__))
+                for cred in self.credentials:
+                    fill(self.credentials[cred].form, self.credentials[cred],
+                         validate=validate_credentials)
+                self._submit(cancel, self.add_provider_button)
+                if not cancel:
+                    flash.assert_message_match('{} Providers '
+                                               '"{}" was saved'.format(self.string_name, self.name))
+            if validate_inventory:
+                self.validate()
+            return created
+
+    def update(self, updates, endpoints=None, cancel=False, validate_credentials=True):
         """
         Updates a provider in the UI.  Better to use utils.update.update context
         manager than call this directly.
 
         Args:
            updates (dict): fields that are changing.
+           endpoints (list of Endpoints): objects which hold information about provider's endpoints
```

Comment:
So after this gets merged, we could get this to be a dict, but if we want to enforce the order, we could keep both for some cases ...